### PR TITLE
Update lastModified timestamp of project on several types of changes

### PR DESCRIPTION
**Pull request by wow-such-code**
* Adds new ProjectChanged event for use cases that make changes (**creation, update, deletion**) to elements of a project (e.g. **samples, batches, measurements, uploaded attachments, experiments**)
* Adds directive **UpdateProjectLastModified** that is executed when this Event happens
* Adds method to change the project modified timestamp (and only that) that can be called by the JobRunner and thus does not need credentials
* Fixes several small issues, typos and inconsistencies related to Event handling

### CODE REVIEW

File: project-management/src/main/java/life/qbic/projectmanagement/application/policy/directive/UpdateProjectUponDeletionEvent.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  public Class<? extends DomainEvent> subscribedToEventType() {
+    return ProjectChanged.class;
+  }
+
+  @Override
+  public void handleEvent(ProjectChanged event) {
+    jobScheduler.enqueue(() -> updateProjectModified(event.projectId(), event.occurredOn()));
+  }
+
+  @Job(name = "Update_Project_Modified")
```
[the remark this answers is not in the document]
Comment (KochTobi) [reply project-management/src/main/java/life/qbic/projectmanagement/application/policy/directive/UpdateProjectUponDeletionEvent.java:45]: The same goes for the other directives.

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java [hunk 2]
```diff
@@ -37,56 +35,20 @@ public MeasurementDomainService(MeasurementRepository measurementRepository) {
...
-    try {
-      measurementRepository.updateNGS(measurement);
-      return Result.fromValue(measurement);
-    } catch (RuntimeException e) {
-      log.error("Measurement update: Failed for measurement with id " + measurement.measurementId()
-          .value(), e);
-    }
-    return Result.fromError(ResponseCode.FAILED);
-  }
-
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java:89]: why was this removed?

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java [hunk 3]
```diff
@@ -70,25 +92,30 @@ public void updateSamples(Project project, Collection<SampleUpdateRequest> updat
...
     DomainEventDispatcher.instance().dispatch(sampleDeleted);
   }
 
-  private void dispatchSuccessfulSampleRegistration(Sample sample) {
-    SampleRegistered sampleRegistered = SampleRegistered.create(sample.assignedBatch(),
-        sample.sampleId());
-    DomainEventDispatcher.instance().dispatch(sampleRegistered);
+  private void dispatchProjectChangedUponSampleDeletion(ProjectId projectId) {
+    ProjectChanged projectChanged = ProjectChanged.create(projectId);
+    DomainEventDispatcher.instance().dispatch(projectChanged);
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java:118]: where did the sample registered event go to?

File: project-management/src/test/groovy/life/qbic/projectmanagement/application/policy/SampleDeletedPolicySpec.groovy [hunk 4]
```diff
@@ -17,13 +19,15 @@ import spock.lang.Specification
 
 class SampleDeletedPolicySpec extends Specification {
 
-    def "Given a sample deletion event, the directive to remove the sample from a batch is executed"() {
+    def "Given a sample deletion event, the respective directive is executed"() {
```
Comment (KochTobi) [inline project-management/src/test/groovy/life/qbic/projectmanagement/application/policy/SampleDeletedPolicySpec.groovy:22]: why the name change?
Maybe it would make sense to test them separately so they can fail separately.

File: project-management/src/test/groovy/life/qbic/projectmanagement/domain/service/SampleDomainServiceSpec.groovy [hunk 5]
```diff
@@ -57,10 +57,12 @@ class SampleDomainServiceSpec extends Specification {
...
-        Result<Collection<Sample>, SampleDomainService.ResponseCode> result = sampleDomainService.registerSamples(project, sampleCodesToRegistrationRequests)
+        sampleDomainService.registerSamples(project, sampleCodesToRegistrationRequests)
 
         then:
-        sampleRegistered.batchIdOfEvent.equals(result.getValue()[0].assignedBatch())
-        sampleRegistered.sampleIdOfEvent.equals(result.getValue()[0].sampleId())
+
+        sampleRegistered.batchIdOfEvent.equals(sampleRegistrationRequest.assignedBatch())
+        SampleId.parse(sampleRegistered.sampleIdOfEvent.value())
+        notThrown(IllegalArgumentException)
```
Comment (KochTobi) [inline project-management/src/test/groovy/life/qbic/projectmanagement/domain/service/SampleDomainServiceSpec.groovy:66]: what changed here?

File: user-interface/src/main/java/life/qbic/datamanager/AppConfig.java [hunk 6]
```diff
@@ -198,9 +216,14 @@ public ProjectRegisteredPolicy projectRegisteredPolicy(SampleCodeService sampleC
 
   @Bean
   public SampleRegisteredPolicy sampleRegisteredPolicy(
-      BatchRegistrationService batchRegistrationService, JobScheduler jobScheduler) {
+      BatchRegistrationService batchRegistrationService,
+      SampleInformationService sampleInformationService,
+      ExperimentInformationService experimentInformationService,
+      ProjectInformationService projectInformationService, JobScheduler jobScheduler) {
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/AppConfig.java:222]: why so many services?

**Review by KochTobi — CHANGES_REQUESTED** [review]

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/measurement/NGSMeasurementJpaRepo.java [hunk 7]
```diff
@@ -21,4 +21,7 @@ public interface NGSMeasurementJpaRepo
 
   Optional<NGSMeasurement> findNGSMeasurementByMeasurementCode(
       MeasurementCode measurementCode);
+
+  Optional<NGSMeasurement> findNGSMeasurementByMeasurementId(
+      MeasurementId measurementId);
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/measurement/NGSMeasurementJpaRepo.java:26]: Many events don't have access to the relevant project ID, so it has to be found in the resulting directive. The directives use different services to fetch the project ID and update it. For measurements in specific there was confusion about which methods fetch by "measurement code" and which by uuid, hence I had to add some methods.

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java (hunk 3 above)
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java:118]: this is the delete event and it cannot be handled inside the Sample object, so @sven1103 and I agreed that we can dispatch a project changed event here, instead.

File: user-interface/src/main/java/life/qbic/datamanager/AppConfig.java (hunk 6 above)
Comment (wow-such-code) [reply user-interface/src/main/java/life/qbic/datamanager/AppConfig.java:222]: batch service to add the sample to the batch,
sample and experiment service to find the project id, which is not connected to the sample,
project service to update the project

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/QbicExperimentRepo.java [hunk 8]
```diff
@@ -12,4 +14,7 @@ public interface QbicExperimentRepo extends JpaRepository<Experiment, Experiment
 
   List<Experiment> findExperimentByExperimentId(ExperimentId id);
 
+  @Query(value = "SELECT project FROM experiments_datamanager WHERE id = ?1",
+      nativeQuery = true)
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/QbicExperimentRepo.java:18]: because due to the current table structure this is not as straightforward as the other queries and we didn't want to make big changes to the experiment or project table

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java (hunk 3 above)
Comment (KochTobi) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java:118]: yes but the sample registered event vanished on your branch. Maybe I am missing something?

File: domain-concept/src/main/java/life/qbic/domain/concepts/DomainEvent.java [hunk 9]
```diff
@@ -1,5 +1,6 @@
 package life.qbic.domain.concepts;
 
+import com.fasterxml.jackson.annotation.JsonGetter;
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply domain-concept/src/main/java/life/qbic/domain/concepts/DomainEvent.java:3]: this helps with serialisation of methods that don't follow the getProperty() notation, in this case: occurredOn()

File: domain-concept/src/main/java/life/qbic/domain/concepts/DomainEvent.java [hunk 10]
```diff
@@ -13,10 +14,21 @@
  */
 public abstract class DomainEvent implements Serializable {
 
+  protected final Instant occurredOn;
+
+  protected DomainEvent() {
+    this.occurredOn = Instant.now();
+  }
+
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply domain-concept/src/main/java/life/qbic/domain/concepts/DomainEvent.java:22]: It IS only fired in the process of instantiation, when super() is called. Since it is used in dozens of subclasses (every one), I think this provides both clarity, as well as ease of refactoring. Every event has a timestamp when it occured.

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java [hunk 11]
```diff
@@ -42,21 +48,37 @@ public SampleDomainService(SampleRepository sampleRepository) {
...
     sampleCodesToRegistrationRequests.forEach((sampleCode, sampleRegistrationRequest) -> {
       var sample = Sample.create(sampleCode, sampleRegistrationRequest);
       samplesToRegister.add(sample);
     });
     Result<Collection<Sample>, ResponseCode> result = this.sampleRepository.addAll(project,
         samplesToRegister);
-    result.onValue(
-            createdSamples -> createdSamples.forEach(this::dispatchSuccessfulSampleRegistration))
-        .onError(Result::fromError);
+    if(result.isValue()) {
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java:65]: because onValue expects me to do something with the sample (result) objects, which are not needed here at all. I don't have strong feelings about this, though, shall I change it anyway?

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java (hunk 3 above)
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/service/SampleDomainService.java:118]: The creation and update events are in the Sample (etc.) classes themselves, now. in the SampleDomainService these events are cached and then forwarded, if registration/update succeeds. In this case in the registerSamples() method.

File: project-management/src/test/groovy/life/qbic/projectmanagement/application/policy/SampleDeletedPolicySpec.groovy (hunk 4 above)
Comment (wow-such-code) [reply project-management/src/test/groovy/life/qbic/projectmanagement/application/policy/SampleDeletedPolicySpec.groovy]: good point, I changed the name and added another test class for project changed events

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/project/ProjectRepositoryImpl.java [hunk 12]
```diff
@@ -99,6 +100,19 @@ public Optional<Project> find(ProjectId projectId) {
...
   }
 
+  /**
+   * Updates the lastModified time of the project. Does not check credentials, as the jobrunner
+   * needs to call it. Use with care.
+   * @param projectId  the id of the project to update
+   * @param modifiedOn the Instant object denoting the time the project was updated
+   */
+  @Override
+  public void updateLastModified(ProjectId projectId, Instant modifiedOn) {
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/project/ProjectRepositoryImpl.java:112]: I was not happy about this, either, but we didn't think there was a better way. I appreciate the suggestion of renaming the method and will do that.

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java [hunk 13]
```diff
@@ -161,95 +132,14 @@ public Collection<NGSMeasurement> findNGSMeasurements(String filter, ExperimentI
 
   @PostAuthorize(
       "hasPermission(#projectId, 'life.qbic.projectmanagement.domain.model.project.Project', 'READ') ")
-  public Collection<NGSMeasurement> findNGSMeasurements(ExperimentId experimentId,
-      ProjectId projectId) {
+  public Collection<NGSMeasurement> findNGSMeasurements(ExperimentId experimentId, ProjectId projectId) {
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java:135]: I think I accidentally removed the ProjectId since it was "not used" - but it's used for authorization :D

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java [hunk 14]
```diff
@@ -161,95 +132,14 @@ public Collection<NGSMeasurement> findNGSMeasurements(String filter, ExperimentI
...
-
-    measurementToUpdate.setMethod(method);
-    measurementToUpdate.setComment(metadata.comment());
-    var updateResult = measurementDomainService.updateNGS(measurementToUpdate);
-
-    if (updateResult.isError()) {
-      return Result.fromError(ErrorCode.FAILED);
-    } else {
-      return Result.fromValue(updateResult.getValue().measurementId());
-    }
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java:252]: these were unused methods that were replaced a few PRs ago by methods handling multiple measurements at once

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java [hunk 15]
```diff
@@ -560,9 +469,28 @@ public CompletableFuture<List<Result<MeasurementId, ErrorCode>>> updateAll(
     } catch (MeasurementRegistrationException e) {
       return CompletableFuture.completedFuture(List.of(Result.fromError(e.reason)));
     }
+    // if the update worked, we forward the unique update events, otherwise it will be rolled back
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java:472]: I mean the error case is returned and the events are not forwarded - I'm not sure if and how the rollback for measurement registration itself was implemented

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java [hunk 16]
```diff
@@ -770,24 +698,6 @@ private NGSMeasurement prepareNGSMeasurementUpdate(NGSMeasurementMetadata metada
...
-        .map(MeasurementService::mergeNGS).filter(Optional::isPresent).map(Optional::get).toList();
-
-    var singleMeasurements = ngsMeasurementMetadataList.stream()
-        .filter(metadata -> metadata.assignedSamplePoolGroup().isEmpty()).toList();
-
-    return Stream.concat(singleMeasurements.stream(),
-        mergedPooledMeasurements.stream()).toList();
-  }
-
-
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java:790]: not used any more

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/sample/event/BatchRegistered.java [hunk 17]
```diff
@@ -1,7 +1,7 @@
 package life.qbic.projectmanagement.domain.model.sample.event;
 
+import com.fasterxml.jackson.annotation.JsonGetter;
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/sample/event/BatchRegistered.java:3]: I think it was an oversight before - it was explained to me that annotating the getters in this way makes job logging more readable

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java [hunk 18]
```diff
@@ -37,56 +35,20 @@ public MeasurementDomainService(MeasurementRepository measurementRepository) {
...
-    try {
-      measurementRepository.save(measurement, sampleCodes);
-      return Result.fromValue(measurement);
-    } catch (RuntimeException e) {
-      log.error(
-          "Saving the NGS measurement failed for id: " + measurement.measurementCode().value(), e);
-    }
-    return Result.fromError(ResponseCode.FAILED);
-  }
-
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java:51]: add methods for single measurements have been replaced by methods for adding multiple measurements and this code was unused.

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java [hunk 19]
```diff
@@ -37,56 +35,20 @@ public MeasurementDomainService(MeasurementRepository measurementRepository) {
...
-      measurementRepository.save(measurement, sampleCodes);
-      return Result.fromValue(measurement);
-    } catch (RuntimeException e) {
-      log.error(
-          "Saving the Proteomics measurement failed for id: " + measurement.measurementCode()
-              .value(), e);
-    }
-    return Result.fromError(ResponseCode.FAILED);
-  }
-
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java:71]: see above

File: project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java (hunk 2 above)
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/service/MeasurementDomainService.java:89]: same as for add methods - replaced by method for updating multiple measurements

File: project-management/src/main/java/life/qbic/projectmanagement/application/api/ProjectPurchaseStorage.java [hunk 20]
```diff
@@ -32,4 +32,5 @@ public interface ProjectPurchaseStorage {
 
   Optional<Offer> findOfferForProject(String projectId, Long offerId);
 
+  Optional<ServicePurchase> findPurchase(Long purchaseID);
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/application/api/ProjectPurchaseStorage.java:35]: they are different objects, although I'm not 100% sure about the domain knowledge behind that. ServicePurchase is closer linked to the uploads and thus what was used here for the events.

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/experiment/Experiment.java [hunk 21]
```diff
@@ -340,5 +344,16 @@ public void setAnalytes(
...
+  }
+
+  private void emitUpdatedEvent() {
+    var updatedEvent = new ExperimentUpdatedEvent(this.experimentId());
+    LocalDomainEventDispatcher.instance().dispatch(updatedEvent);
+  }
+
+  private void emitCreatedEvent() {
+    var createdEvent = new ExperimentCreatedEvent(this.experimentId());
+    LocalDomainEventDispatcher.instance().dispatch(createdEvent);
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/experiment/Experiment.java:357]: so for now we focused on add, edit and delete

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/experiment/Experiment.java [hunk 22]
```diff
@@ -95,27 +104,6 @@ public List<ExperimentalVariable> variables() {
...
-   * @return a {@link Result} object containing the added level value or declarative exceptions. The
-   * result will contain an {@link ExperimentalVariableNotDefinedException} if no variable with the
-   * provided name is defined in this design.
-   * @since 1.0.0
-   */
-  public Result<VariableLevel, Exception> addLevelToVariable(String variableName,
-      ExperimentalValue level) {
-    return experimentalDesign.addLevelToVariable(variableName, level);
-  }
-
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/experiment/Experiment.java:118]: not used, although now I'm not sure if we should keep it, as it's public and might be used in the future?

**Review by KochTobi — APPROVED** [review]
